### PR TITLE
Type=notify and TimeoutStartSec to wait for ready when starting 

### LINKED ISSUE #8220: Add facultative forking mode or systemd-notify at mimir start
#### Is your feature request related to a problem? Please describe.

When (re)starting Mimir with systemd, the `systemctl` command is always successful even in case of faulty config or external problem (no space left, not enough ram, etc) because mimir service is in `simple` mode, which means that systemd job is only to start the process without waiting for it to fork or sending a notification to systemd that initialization went well. This is problematic because when restarting mimir on a large amount of instances with Ansible for instance, the playbook run should show an error instead showing all good.

#### Describe the solution you'd like

Any of those 2 solutions should work:
* Have a Mimir setting to make it fork after initialization (config parsed, sockets opened, s3 connection opened, etc...) that would make possible to use `type=forking` systemd service.
* Implement systemd-notify protocol to announce to systemd that it's ready, in that case we could use a `type=notify` systemd service.

#### Describe alternatives you've considered

Use a `ExecStartPost` directive in the service containing a sleep then check for the process health, or do that in the Ansible playbook instead. This workaround isn't ideal because time-based sanity checks are not the more reliable in this context in my opinion.

### PR DESCRIPTION
**Pull request by pbailhache**
#### What this PR does

Following #8541 this activate the necessary settings to use the sd_notify event.

#### Which issue(s) this PR fixes or relates to

Fixes #8220 

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by pbailhache** [comment]
Waiting takes a bit more time at startup of mimir :
On an ingester :
before 
```
time systemctl start mimir.service 

real	0m0.021s
user	0m0.007s
sys	0m0.013s
```

after
```
time systemctl start mimir.service 

real	0m0.369s
user	0m0.008s
sys	0m0.016s
```

On a compactor :
before
```
time systemctl start mimir.service 

real	0m0.017s
user	0m0.007s
sys	0m0.007s
```

after
```
time systemctl start mimir.service 

real	1m1.222s
user	0m0.015s
sys	0m0.001s
```

This will reduce the needs to poll /ready while starting up systemd service.

File: packaging/nfpm/mimir/mimir.service [hunk 1]
```diff
@@ -11,6 +11,8 @@ After=network-online.target
 
 [Service]
 Restart=always
+Type=notify
+TimeoutStartSec=3min
```
Comment (wilfriedroset) [inline packaging/nfpm/mimir/mimir.service:15]: good job. for maintainers here is the doc:
* https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Type=
* https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#TimeoutStartSec=

**Review by wilfriedroset — APPROVED** [review]

**Review by 56quarters — APPROVED** [review]
Nicely done, thank you!

File: packaging/nfpm/mimir/mimir.service (hunk 1 above)
Comment (dimitarvdimitrov) [reply packaging/nfpm/mimir/mimir.service:15]: 3 minutes looks a bit tight to me. Some components will spend a minute just waiting for others to come back up. New store-gateways for example need to then download index headers, which can easily exceed 10 minutes. Can we set this to 20min maybe?

**Comment by ying-jeanne** [comment]
Hi, not familiar with this, @pbailhache could you elaborate more on why TimeoutStartSec is needed and how 3 mins is calculated?

**Comment by pbailhache** [comment]
> Hi, not familiar with this, @pbailhache could you elaborate more on why TimeoutStartSec is needed and how 3 mins is calculated?

Yes, of couse !

This is an improvement over the current systemd configuration.

Currently when mimir starts, it launch all services in an async way. The systemd service state is "active" directly and you need to poll /ready to check if mimir is really started.

Following this [PR](https://github.com/grafana/mimir/pull/8541/commits/e97ba0b4fa0d3a6b3b0cb7a2a7bcd536382906ff#diff-d1a605160f2b763dbf8d1feea0506ac9ee143a51b6a303815cf56b411db44c4c), we use [sd_notify](https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html) to get a notification when the services are started. 

This notification is then used to [notify](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Type=) systemd about the mimir.service started status.

[TimeoutStartSec](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#TimeoutStartSec=) define the max time waiting for start before considering that the service is failed.

3 min is a value we chose based on our max starting time its a rough value and as @dimitarvdimitrov said maybe we need to increase it. 20 min is ok on my side but this also means that the service won't be considered "failed" until the end of this timeout (and the systemd start/restart call will hang until then).

**Comment by ying-jeanne** [comment]
> > Hi, not familiar with this, @pbailhache could you elaborate more on why TimeoutStartSec is needed and how 3 mins is calculated?
> 
> Yes, of couse !
> 
> This is an improvement over the current systemd configuration.
> 
> Currently when mimir starts, it launch all services in an async way. The systemd service state is "active" directly and you need to poll /ready to check if mimir is really started.
> 
> Following this [PR](https://github.com/grafana/mimir/pull/8541/commits/e97ba0b4fa0d3a6b3b0cb7a2a7bcd536382906ff#diff-d1a605160f2b763dbf8d1feea0506ac9ee143a51b6a303815cf56b411db44c4c), we use [sd_notify](https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html) to get a notification when the services are started.
> 
> This notification is then used to [notify](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Type=) systemd about the mimir.service started status.
> 
> [TimeoutStartSec](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#TimeoutStartSec=) define the max time waiting for start before considering that the service is failed.
> 
> 3 min is a value we chose based on our max starting time its a rough value and as @dimitarvdimitrov said maybe we need to increase it. 20 min is ok on my side but this also means that the service won't be considered "failed" until the end of this timeout (and the systemd start/restart call will hang until then).

The problem is even we can estimate the max starting time by your test, the real start time is depending on the datas user ingested, how about set this with environment variable instead of hardcoded value.

File: packaging/nfpm/mimir/mimir.service (hunk 1 above)
Comment (dimitarvdimitrov) [reply packaging/nfpm/mimir/mimir.service:15]: or if the setting is not required maybe we can unset it? We deploy Mimir without a liveness probe in k8s and it generally must always either start or fail, but never get stuck.

**Comment by wilfriedroset** [comment]
> how about set this with environment variable instead of hardcoded value.

I would suggest that we rely on `systemd drop-in` feature which is made exactly for this type of usecases.
See: https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#Examples

following this path, users should create an override file in the correct path similar to this 

```ini
cat /etc/systemd/system/mimir.service.d/mimir-dropin.conf
# MANAGED BY PUPPET
[Service]
LimitNOFILE=1048576
# Overriding vendor settings because mimir takes more time than the default value. 
TimeoutStartSec=30min 
```

I believe it fits nicely in systemd design pattern.
The alternative would be to create a file in /etc/default/mimir (see: the current file https://github.com/grafana/mimir/blob/main/packaging/nfpm/mimir/mimir.env), add an environment variable and then set `TimeoutStartSec=$TimeoutStartSec`

This adds an extra layer of indirection which I find not user friendly and hide the real information, especially when you use `systemctl` command
Example:

```
mimir-[redacted] ~ # systemctl cat mimir
# /lib/systemd/system/mimir.service
# SPDX-License-Identifier: AGPL-3.0-only
# Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/packaging/deb/systemd/cortex.service
# Provenance-includes-license: Apache-2.0
# Provenance-includes-copyright: The Cortex Authors.

[Unit]
Description=Horizontally scalable, highly available, multi-tenant, long term Prometheus.
Documentation=https://grafana.com/oss/mimir/
Wants=network-online.target
After=network-online.target

[Service]
Restart=always
User=mimir
EnvironmentFile=/etc/default/mimir
ExecStart=/usr/local/bin/mimir --config.file=/etc/mimir/config.yml --runtime-config.file=/etc/mimir/runtime_config.yml --log.level $LOG_LEVEL $CUSTOM_ARGS
TimeoutStopSec=20s
SendSIGKILL=no
WorkingDirectory=/var/lib/mimir

[Install]
WantedBy=multi-user.target

# /etc/systemd/system/mimir.service.d/mimir-dropin.conf
# MANAGED BY PUPPET
[Service]
LimitNOFILE=1048576
Environment=JAEGER_SERVICE_NAME=[redacted] JAEGER_ENDPOINT=https://[redacted]:14268/api/traces JAEGER_SAMPLER_TYPE=probabilistic JAEGER_SAMPLER_PARAM=0.01 JAEGER_TAGS=cluster_id=1>
StandardOutput=append:/var/log/mimir/mimir.log
StandardError=inherit
[Unit]
After=consul.service
```
```
mimir-[redacted] ~ # systemctl show mimir.service
Type=simple
Restart=always
NotifyAccess=none
RestartUSec=100ms
TimeoutStartUSec=1min 30s
TimeoutStopUSec=20s
TimeoutAbortUSec=20s
TimeoutStartFailureMode=terminate
TimeoutStopFailureMode=terminate
RuntimeMaxUSec=infinity
WatchdogUSec=0
WatchdogTimestampMonotonic=0
RootDirectoryStartOnly=no
--8<--
```

wdyt @ying-jeanne ?

File: packaging/nfpm/mimir/mimir.service (hunk 1 above)
Comment (wilfriedroset) [reply packaging/nfpm/mimir/mimir.service:15]: I'm in favor of `20min` by default.
as pointed by @pbailhache the command `systemctl start mimir` will hang for `TimeoutStartSec` value.

**Comment by ying-jeanne** [comment]
> > how about set this with environment variable instead of hardcoded value.
> 
> I would suggest that we rely on `systemd drop-in` feature which is made exactly for this type of usecases. See: https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#Examples
> 
> following this path, users should create an override file in the correct path similar to this
> 
> ```ini
> cat /etc/systemd/system/mimir.service.d/mimir-dropin.conf
> # MANAGED BY PUPPET
> [Service]
> LimitNOFILE=1048576
> # Overriding vendor settings because mimir takes more time than the default value. 
> TimeoutStartSec=30min 
> ```
> 
> I believe it fits nicely in systemd design pattern. The alternative would be to create a file in /etc/default/mimir (see: the current file https://github.com/grafana/mimir/blob/main/packaging/nfpm/mimir/mimir.env), add an environment variable and then set `TimeoutStartSec=$TimeoutStartSec`
> 
> This adds an extra layer of indirection which I find not user friendly and hide the real information, especially when you use `systemctl` command Example:
> 
> ```
> mimir-[redacted] ~ # systemctl cat mimir
> # /lib/systemd/system/mimir.service
> # SPDX-License-Identifier: AGPL-3.0-only
> # Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/packaging/deb/systemd/cortex.service
> # Provenance-includes-license: Apache-2.0
> # Provenance-includes-copyright: The Cortex Authors.
> 
> [Unit]
> Description=Horizontally scalable, highly available, multi-tenant, long term Prometheus.
> Documentation=https://grafana.com/oss/mimir/
> Wants=network-online.target
> After=network-online.target
> 
> [Service]
> Restart=always
> User=mimir
> EnvironmentFile=/etc/default/mimir
> ExecStart=/usr/local/bin/mimir --config.file=/etc/mimir/config.yml --runtime-config.file=/etc/mimir/runtime_config.yml --log.level $LOG_LEVEL $CUSTOM_ARGS
> TimeoutStopSec=20s
> SendSIGKILL=no
> WorkingDirectory=/var/lib/mimir
> 
> [Install]
> WantedBy=multi-user.target
> 
> # /etc/systemd/system/mimir.service.d/mimir-dropin.conf
> # MANAGED BY PUPPET
> [Service]
> LimitNOFILE=1048576
> Environment=JAEGER_SERVICE_NAME=[redacted] JAEGER_ENDPOINT=https://[redacted]:14268/api/traces JAEGER_SAMPLER_TYPE=probabilistic JAEGER_SAMPLER_PARAM=0.01 JAEGER_TAGS=cluster_id=1>
> StandardOutput=append:/var/log/mimir/mimir.log
> StandardError=inherit
> [Unit]
> After=consul.service
> ```
> 
> ```
> mimir-[redacted] ~ # systemctl show mimir.service
> Type=simple
> Restart=always
> NotifyAccess=none
> RestartUSec=100ms
> TimeoutStartUSec=1min 30s
> TimeoutStopUSec=20s
> TimeoutAbortUSec=20s
> TimeoutStartFailureMode=terminate
> TimeoutStopFailureMode=terminate
> RuntimeMaxUSec=infinity
> WatchdogUSec=0
> WatchdogTimestampMonotonic=0
> RootDirectoryStartOnly=no
> --8<--
> ```
> 
> wdyt @ying-jeanne ?
> 
> ps: here is another example in real life about using drop-in https://docs.percona.com/percona-xtradb-cluster/5.7/management/configure-timeoutstartsec.html

agree, drop-in feature is quite nit.

**Review by ying-jeanne — APPROVED** [review]
LGTM

**Comment by dimitarvdimitrov** [comment]
I still think the startup timeout of 3m is too low: opened https://github.com/grafana/mimir/pull/8658 as a follow-up.

**Pull request merged**